pkgs/applications/science/math/sage/wrap.nix

```diff
@@ -0,0 +1,48 @@
+{ lib
+, stdenv
+, writeTextFile
+, python
+}:
+
+{ packageToWrap
+, module_name ? packageToWrap.pname
+, set ? {}
+, prefix ? {}
+, code ? ""
+, extraInstall ? ""
+}:
+
+let
+  pythonEscape = str: lib.replaceStrings ["'"] ["a"] (toString str);
+  generateExport = var: val: "os.environ['${pythonEscape var}'] = '${pythonEscape val}'";
+  generatePrefix = var: val: "os.environ['${pythonEscape var}'] = '${pythonEscape val}:' + os.environ.get('${pythonEscape var}', '')";
+  generateWrapperLines = { set, prefix}:
+    (lib.mapAttrsToList generateExport set)
+      ++ (lib.mapAttrsToList generatePrefix prefix);
+  scriptToExecute = "import os\n" + lib.concatStringsSep "\n" (generateWrapperLines { inherit set prefix; }) + "\n" + code;
+  targetDir = "${placeholder "out"}/lib/${python.libPrefix}/site-packages/${module_name}/";
+  wrapperScript = ''
+    import sys
+    exec(compile(open('${targetDir}/inject.py').read(), '${targetDir}/inject.py', 'exec'), {}, {})
+    sys.path = [ '${packageToWrap}/lib/${python.libPrefix}/site-packages' ] + sys.path
+    sys.modules.pop(__name__)
+    import ${module_name}
```

Comment:
@FRidh @dotlambda what do you think about this?

It turns out we actually *can* wrap python libraries (continuing the discussion in https://github.com/NixOS/nixpkgs/pull/53816). The trick is to create a module with the same name that does whatever the wrapper is supposed to do and then swaps itself out with the actual module (like `exec` in bash).

Here I use this for sage, but it could be generalized and used for gnome etc. where we need to set certain environment variables.

There is a usage example in `sage-env.nix` (better look at the file directly instead of the diff).